File: core/lib/template.php
<?php
###########################################################
# Copyright 2011 Daniel 'grindhold' Brendle and Team
#
# This file is part of Scoville.
#
# Scoville is free software: you can redistribute it and/or 
# modify it under the terms of the GNU General Public License 
# as published by the Free Software Foundation, either 
# version 3 of the License, or (at your option) any later 
# version.
#
# Scoville is distributed in the hope that it will be 
# useful, but WITHOUT ANY WARRANTY; without even the implied 
# warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR 
# PURPOSE. See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public 
# License along with Scoville. 
# If not, see http://www.gnu.org/licenses/.
###########################################################
require_once 'core.php';

class TemplateException extends Exception{}

class TemplateManager extends Singleton{
	private static $instance = null;
	
	/**
	 * Get Singleton Instance
	 * 
	 * Returns the singleton Instance of the Rightsmanager
	 * 
	 * @return RightsManager The rights manager
	 */
	public static function getInstance(){
		if (TemplateManager::$instance==null){
			TemplateManager::$instance = new TemplateManager();
			TemplateManager::$instance->init();
		}
		return TemplateManager::$instance;
	}
	
	protected function init(){}
	
	public function createFromData($data){
		$template = new Template();
		$filename =hash('md5',time()+session_id()).".tar.gz";
		$template->setFilename($filename); 
		$handle = fopen("/tmp/".$filename,"w");
		fwrite($handle,$data);
		fclose();
		return $template;
	}
	
	public function createFromFile($filename){
		$template = new Template();
		$template->setFilename($filename);
		return $template;
	}
	
	public function createFromRemote($repository,$templatename){
		$template = $repository->downloadTemplate($templatename);
		$data = $template->data;
		return $this->createFromData($data);
	}
	
	public function createCurrentInstalled(){
		global $SCV_GLOBALCFG;
		$core = Core::getInstance();
		$template = new Template();
		$manifestWebPath = $SCV_GLOBALCFG['SCV_WEBPATH'].$SCV_GLOBALCFG['SCV_INSTANCE_SCOPE_ID']."/web/manifest.json";
		$manifestRaw = file_get_contents($manifestWebPath);
		if ($manifestRaw == false){
			throw new ModuleException("InstallationError: The manifest is not a valid Scoville Template");
			return;
		}
		$manifest = json_decode($manifestRaw);
		if ($manifest == null){
			throw new ModuleException("InstallationError: Manifest seems to be broken. Validate!");
			return;
		}
		$template->setManifest($manifest);
		$template->setInstalled(true);
		return $template;
	}
}

class Template {
	private $manifest = null;
	private $filename = null;
	private $installed = false;
	
	public function uninstall($tryToMap=false, $checkRight=true){
		if(!$this->installed){
			throw new TemplateException("Uninstall: Cannot uninstall a non-installed Template");
		}
		foreach($this->manifest->filenames as $filename){
			unlink("../web/".$filename);
		}
		
		$core = Core::getInstance();
		$db = $core->getDB();
		
		$sitesToSelect = array();
		
		foreach ($this->manifest->sites as $site){
			if (preg_match('/^[\w_-]*.html$/',$site->filename)){
				$sitesToSelect[] = $site->filename;
			}else{
				error_log("TemplateSystem//Warning: Site has been ignored due to suspicious filename: ".$site->filename);
			}
		}
		
		$stmnt = "SELECT SIT_ID FROM SITES WHERE SIT_FILENAME IN ( '".join("','",$sitesToSelect)."' ) ;";
	    $res = $db->query($core,$stmnt);
		
		$siteIds = array();
		while($set = $db->fetchArray($res)){
			$siteIds[] = $set['SIT_ID'];
		}
		if (count($siteIds)>0){
			if (!$tryToMap){
				$stmnt = "UPDATE WIDGETS SET WGT_SPACE = NULL WHERE WGT_SIT_ID IN (".join(",",$siteIds).") ;";
				$db->query($core,$stmnt);
			}
			$stmnt = "DELETE FROM SITES WHERE SIT_ID IN (".join(",",$siteIds).") ;";
			$db->query($core,$stmnt);
			$stmnt = "UPDATE WIDGETS SET WGT_SIT_ID = NULL WHERE WGT_SIT_ID IN (".join(",",$siteIds).") ;";
			$db->query($core,$stmnt);
		}
		$manifestWebPath = $SCV_GLOBALCFG['SCV_WEBPATH'].$SCV_GLOBALCFG['SCV_INSTANCE_SCOPE_ID']."/web/manifest.json";
		unlink($manifestWebPath);
	}
	
	
	public function validate(){
		//TODO: Eventuell noch nach im template vorhandenen ordnern ueberpruefen, die dort nicht sein sollten.
		
		if ($this->installed){
			throw new TemplateException("Validate: Cannot validate the installed template");
		}
		if (!isset($this->manifest)){
			throw new TemplateException("Validate: Cannot validate without a manifest");
		}
		if (!is_dir("/tmp/".$this->getTemporaryFolder())){
			throw new TemplateException("Validate: Cannot validate a Template that is not temporarily unpacked");
		}
		
		$foundDefault = false;
		
		foreach($this->manifest->sites as $site){
			if (isset($site->default)){
				$foundDefault = true;
			}
			if (!file_exists("/tmp/".$this->getTemporaryFolder()."/".$site->filename)){
				throw new TemplateException("Validate: File does not Exist: $site->filename");
			}
			$sitename = str_replace(".html","",$site->filename);
			if (!file_exists("/tmp/".$this->getTemporaryFolder()."/".$sitename.".css")){
				throw new TemplateException("Validate: Site is missing CSS: $site->filename");
			}
		}
		
		if(!file_exists("/tmp/".$this->getTemporaryFolder()."/general.css")){
			throw new TemplateException("Validate: Could not find general.css");
		}
		
		if (!$foundDefault){
			throw new TemplateException("Validate: Could not find Defaultsite");
		}
		
	}
	
	public function getSiteSpaces($sitedata){
		$spaceId = 1;
		while(true){
			if (1 == preg_match('/<\s*div[^>]*id\s*=\s*"s'.$spaceId.'"[^>]*>/',$sitedata)) {
				$spaceId++;
			}else{
				break;
			}
		}
		return $spaceId-1;
	}
	
	public function install($tryToMap=false, $checkRight=true){
		global $SCV_GLOBALCFG;

		if ($this->installed){
			throw new TemplateException("Installation: This template is already installed!");
		}
		
		$core = Core::getInstance();
		$templateM = $core->getTemplateManager();
		$currentTemplate = $templateM->createCurrentInstalled();
		$currentTemplate->uninstall($tryToMap, $checkRight);
		
		$this->temporaryUnpack();
		
		$manifestRaw = file_get_contents("/tmp/".$this->getTemporaryFolder()."/manifest.json");
		if ($manifestRaw == false){
			throw new ModuleException("InstallationError: $moduleId is not a valid Scoville Module");
		}
		$manifest = json_decode($manifestRaw);
		if ($manifest == null){
			throw new ModuleException("InstallationError: Manifest seems to be broken. Validate!");
		}
		$this->setManifest($manifest);
		
		$this->validate();
		
		foreach($this->manifest->filenames as $file){
			copy("/tmp/".$this->getTemporaryFolder()."/".$file, "../web/".$file);
		}
		
		$cssParser = new CssParser(file_get_contents("/tmp/".$this->getTemporaryFolder()."/general.css"));
		$core = Core::getInstance();
		$cssM = $core->getCssManager();
		$serverPropertySet = $cssM->getCssPropertySet();
		$serverPropertySet->setFromParser($cssParser);
		$serverPropertySet->store();
		
		$db = $core->getDB();
		$stmnt = "INSERT INTO SITES (SIT_ID, SIT_NAME, SIT_DESCRIPTION, SIT_FILENAME, SIT_HTML, SIT_SPACES, SIT_DEFAULT ) VALUES (GEN_ID(SIT_GEN,1),?,?,?,?,?,?) ;";
		foreach($this->manifest->sites as $site){
			$sitedata = file_get_contents("../web/".$site->filename);
			$spaces = $this->getSiteSpaces($sitedata);
			$name = isset($site->name)?"(unnamed)":$site->name;
			$htmlBlob = $db->createBlob($sitedata);
			$db->query($core,$stmnt,array($site->name, $site->desc, $site->filename, $htmlBlob, $spaces, isset($site->default)?1:0));
		}
		$manifestWebPath = $SCV_GLOBALCFG['SCV_WEBPATH'].$SCV_GLOBALCFG['SCV_INSTANCE_SCOPE_ID']."/web/manifest.json";
		copy("/tmp/".$this->getTemporaryFolder()."/manifest.json", $manifestWebPath);
		return;
	}
	
	public function setManifest($manifest){
		if (is_object($manifest) and $this->validateManifest($manifest)){
			$this->manifest = $manifest;
		}else{
			throw new TemplateException("Set Manifest: The manifest does not match specification!");
		}
	}
	
	public function getManifest(){
		return $this->manifest;
	}
	
	public function setFilename($filename){
		$this->filename = $filename;
	}
	
	public function setInstalled($installed){
		$this->installed = $installed;
	}
	
	public function temporaryUnpack(){
		if (!is_dir("/tmp/".$this->getTemporaryFolder())){
			mkdir("/tmp/".$this->getTemporaryFolder());
			system('tar xfz /tmp/'.escapeshellarg($this->filename).' -C /tmp/'.escapeshellarg($this->getTemporaryFolder()).'/ > /dev/null');
		}
	}
	
	private function getTemporaryFolder(){
		if (!isset($this->filename)){
			throw new TemplateException("GetTemporaryFolder: Cannot get Filename (Is this template the installed one? ;> )");
		}
		return str_replace(".tar.gz","",$this->filename);
	}
	
	public function __destruct(){
		if (isset($this->filename)){
			if (file_exists("/tmp/".$this->filename)){
				unlink("/tmp/".$this->filename);
			}
			if (is_dir("/tmp/".$this->getTemporaryFolder())){
				$core = Core::getInstance();
				$core->recursiveDelete("/tmp/".$this->getTemporaryFolder());
			}
		}
		
	}
	
	private function validateManifest($manifest){
		if (!isset($manifest->name) or !isset($manifest->description) or !isset($manifest->author)
		       or !isset($manifest->sites) or !isset($manifest->filenames)){
       		return false;
		}
		if (count($manifest->sites) < 1){
			return false;
		}
		foreach ($manifest->sites as $site){
			if (!isset($site->name) or !isset($site->desc) or !isset($site->filename)){
				return false;
			}
		}
		return true;
	}
}
?>

File: core/lib/template.py
#!/usr/bin/python
#-*- coding: utf-8 -*-

###########################################################
# © 2011 Daniel 'grindhold' Brendle and Team
#
# This file is part of Skarphed.
#
# Skarphed is free software: you can redistribute it and/or 
# modify it under the terms of the GNU Affero General Public License 
# as published by the Free Software Foundation, either 
# version 3 of the License, or (at your option) any later 
# version.
#
# Skarphed is distributed in the hope that it will be 
# useful, but WITHOUT ANY WARRANTY; without even the implied 
# warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR 
# PURPOSE. See the GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public 
# License along with Skarphed. 
# If not, see http://www.gnu.org/licenses/.
###########################################################

import os
import tarfile
import shutil
from StringIO import StringIO
from json import JSONDecoder

from skarphedcore.configuration import Configuration
from skarphedcore.database import Database
from skarphedcore.core import Core
from skarphedcore.module import ModuleManager
from skarphedcore.binary import Binary
from skarphedcore.css import CSSManager
from skarphedcore.view import Page, View
from skarphedcore.poke import PokeManager

from common.enums import ActivityType
from common.errors import TemplateException


class Template(object):
    """
    The template class manages the Data that has to be handled with
    template-packages and represents the template to the 

    Furthermore it hold the template-data in RAM to deliver it faster
    to the clients.
    """
    @classmethod
    def get_current_template(cls):
        """
        Receives .tar.gz'ed data and generates templatedata from it
        """
        db = Database()
        stmnt = "SELECT TPL_NAME, TPL_DESC, TPL_AUTHOR FROM TEMPLATE_INFO ;"

        cur = db.query(stmnt)
        tpldata = cur.fetchonemap()
        if tpldata is None:
            raise TemplateException(TemplateException.get_msg(1))

        tpl = Template()
        tpl.set_name(tpldata['TPL_NAME'])
        tpl.set_description(tpldata['TPL_DESC'])
        tpl.set_author(tpldata['TPL_AUTHOR'])

        stmnt = "SELECT TPB_BIN_ID FROM TEMPLATE_BINARIES WHERE TPB_TPL_ID = 0 ;"
        cur = db.query(stmnt)
        rows = cur.fetchallmap()
        for row in rows:
            tpl.add_binary(row['TPB_BIN_ID'])

        return tpl

    @classmethod
    def fetch_templates_for_gui(cls):
        repository = ModuleManager.get_repository()
        data = repository.get_all_templates()
        return data

    @classmethod
    def install_from_repo(cls, nr):
        repository = ModuleManager.get_repository()
        data = repository.download_template(nr)
        return cls.install_from_data(data)

    @classmethod
    def install_from_data(cls, data):
        """
        Receives .tar.gz'ed data and generates templatedata from it
        First validates the data. While validating it tracks all occuring
        errors in the errorlog. If one severe error happens during validation,
        the method stops before actually doing write-operations and returns
        the errorlog to the client
        Otherwise, it executes the installation and returns all 
        non-severe errors (warnings).
        """
        def cleanup(path):
            shutil.rmtree(path)

        #TODO: Mutex this operation

        errorlog = []

        configuration = Configuration()
        webpath = configuration.get_entry("core.webpath")
        temp_installpath = webpath+"/tpl_install"
        os.mkdir(temp_installpath)

        tar = open(temp_installpath+"/tpl.tar.gz","w")
        tar.write(data)
        tar.close()

        tar = tarfile.open(temp_installpath+"/tpl.tar.gz","r:gz")
        tar.extractall(temp_installpath)
        tar.close()

        os.unlink(temp_installpath+"/tpl.tar.gz")

        manifest_file = open(temp_installpath+"/manifest.json","r")
        try:
            manifest = JSONDecoder().decode(manifest_file.read())
        except ValueError,e:
            errorlog.append({'severity':1,
                           'type':'PackageFile',
                           'msg':'JSON seems to be corrupt'})
            cleanup(temp_installpath)
            return errorlog
        manifest_file.close()

        #BEGIN TO VALIDATE DATA

        try:
            f = open(temp_installpath+"/general.css")
            general_css = f.read()
            f.close()
        except IOError,e:
            errorlog.append({'severity':1,
                           'type':'PackageFile',
                           'msg':'File not in Package general.css'})
        
        css_manager = CSSManager()
        general_csspropertyset = None
        try:
            general_csspropertyset = css_manager.create_csspropertyset_from_css(general_css)
            general_csspropertyset.set_type_general()
        except Exception, e:
            errorlog.append({'severity':1,
                           'type':'CSS-Data',
                           'msg':'General CSS File does not Contain Valid CSS '+str(e)})
        
        pagedata = [] # Prepared filedata for execution into Database

        for page in manifest['pages']:
            if page['filename'].endswith(".html"):
                name = page['filename'].replace(".html","",1)
            elif page['filename'].endswith(".htm"):
                name = page['filename'].replace(".htm","",1)
            else:
                errorlog.append({'severity':1,
                               'type':'PageData',
                               'msg':'Invalid format (allowed is .html and .htm: '+page['filename']})

            try:
                f = open(temp_installpath+"/"+page['filename'])
                html = f.read()
                f.close()
            except IOError,e:
                errorlog.append({'severity':1,
                               'type':'PageFile',
                               'msg':'File not in Package '+page['filename']})
                continue

            try:
                f = open(temp_installpath+"/"+name+"_head.html","r")
                html_head = f.read()
                f.close()
            except IOError,e:
                errorlog.append({'severity':0,
                               'type':'PageFile',
                               'msg':'File not in Package '+name+"_head.html"})
                html_head = ""

            try:
                f = open(temp_installpath+"/static/"+name+".css")
                css = f.read()
                f.close()
            except IOError,e:
                errorlog.append({'severity':1,
                               'type':'PageFile',
                               'msg':'File not in Package static/'+name+".css"})
                continue

            try:
                f = open(temp_installpath+"/static/"+name+"_minimap.png","rb")
                minimap = f.read()
                f.close()
                os.unlink(temp_installpath+"/static/"+name+"_minimap.png")
            except IOError,e:
                errorlog.append({'severity':0,
                               'type':'PageFile',
                               'msg':'File not in Package static/'+name+"_minimap.png"})
                minimap = None

            pagedata.append({'name':page['name'],
                             'desc':page['desc'],
                             'html_body':html,
                             'html_head':html_head,
                             'css':css,
                             'minimap':minimap,
                             'internal_name':name})
        
        if len(errorlog) > 0:
            is_severe_error = False
            for error in errorlog:
                if error['severity'] >= 1:
                    is_severe_error = True
                    break
            if is_severe_error:
                cleanup(temp_installpath)
                return errorlog

        # BEGIN TO WRITE DATA

        #release maintenance mode at the end?
        release_maintenance_mode = not cls.is_template_installed()

        #uninstall old template
        if cls.is_template_installed():
            old_template = cls.get_current_template()
            old_template.uninstall()

        new_template = Template()
        new_template.set_name(manifest['name'])
        new_template.set_description(manifest['description'])
        new_template.set_author(manifest['author'])

        #create pages
        for page in pagedata:
            Page.create(page['name'],
                                page['internal_name'],
                                page['desc'],
                                page['html_body'],
                                page['html_head'],
                                page['css'],
                                page['minimap'])    

        #put binary into database
        for bin_filename in os.listdir(temp_installpath+"/static"):
            binary=None
            try:
                bin_file = open(temp_installpath+"/static/"+bin_filename,"rb")
                bin_data = bin_file.read()
                bin_file.close()
                # TODO: Find more generic way to determine mimetype
                if bin_filename.endswith(".png"):
                    binary = Binary.create("image/png", bin_data)
                if bin_filename.endswith(".jpeg") or bin_filename.endswith(".jpg"):
                    binary = Binary.create("image/jpeg", bin_data)
                else:
                    binary = Binary.create("application/octet-stream", bin_data)
                if binary is not None:
                    binary.set_filename(bin_filename)
                    binary.store()

                    new_template.add_binary(binary.get_id())

            except IOError, e:
                errorlog.append({'severity':0,
                               'type':'PageFile',
                               'msg':'File seems broken static/'+bin_filename})


        #read general.css into CSSPropertysets
        general_csspropertyset.store()

        new_template.store()
        cleanup(temp_installpath)

        #create a defaultview if there isnt
        View.create_default_view()

        if release_maintenance_mode:
            Core().deactivate_maintenance_mode()

        return errorlog


    @classmethod
    def is_template_installed(cls):
        """
        checks whether there is a template installed
        """
        db = Database()
        stmnt = "SELECT COUNT(*) AS AMNT FROM TEMPLATE_INFO ;"
        cur = db.query(stmnt)
        row = cur.fetchonemap()
        return bool(row['AMNT'])

    def __init__(self):
        self._name = None
        self._description = None
        self._author = None

        self._binaries = [] #binaries that belong to this template

    def get_id(self):
        """
        everything in skarphed should have a get_id()
        """
        return 0

    def set_name(self, name):
        self._name = unicode(name)

    def get_name(self):
        return self._name 

    def set_description(self, desc):
        self._description = unicode(desc)

    def get_description(self):
        return self._description

    def set_author(self, author):
        self._author = unicode(author)

    def get_author(self):
        return self._author

    def add_binary(self, bin_id):
        if bin_id not in self._binaries:
            self._binaries.append(bin_id)

    def remove_binary(self, bin_id):
        self._binaries.remove(bin_id)

    def store(self):
        """
        stores the template information in the database
        """
        db = Database()
        stmnt = "UPDATE OR INSERT INTO TEMPLATE_INFO (TPL_ID, TPL_NAME, TPL_DESC, TPL_AUTHOR) \
                 VALUES (0, ?, ?, ? ) MATCHING (TPL_ID) ;"
        db.query(stmnt, (self._name, self._description, self._author), commit=True)

        stmnt = "INSERT INTO TEMPLATE_BINARIES (TPB_TPL_ID, TPB_BIN_ID) VALUES (?,?) ;"
        for bin_id in self._binaries:
            db.query(stmnt, (0, bin_id), commit=True)
        PokeManager.add_activity(ActivityType.TEMPLATE)

    def uninstall(self):
        """
        Uninstalls this template
        """
        db = Database()

        for bin_id in self._binaries:
            bin = Binary.get_by_id(bin_id)
            bin.delete()
        stmnt = "DELETE FROM TEMPLATE_BINARIES ;"
        db.query(stmnt, commit=True)
        
        #Destroy Pages
        Page.delete_all_pages()

        #Set Page ID-Generator to 1
        db.set_seq_to('SIT_GEN',1)

        stmnt = "DELETE FROM TEMPLATE_INFO ;"
        db.query(stmnt, commit=True)

        PokeManager.add_activity(ActivityType.TEMPLATE)
